### PR TITLE
Rewrote Manga Info and Manga Sync => new Manga Info (with Anilist support)

### PR DESCRIPTION
**Pull request by NoodleMage**
#312, #340

## TODO

- [x] Comments TODO

- [x] Default Sync Service (in settings)

- [x] Remove old classes

- [x] Test API 16

- [x] Bug fixes

- [x] #200 

## Screens

Feedback is appreciated!

### CODE REVIEW

File: app/src/main/java/eu/kanade/tachiyomi/data/database/models/MangaSyncImpl.kt [hunk 1]
```diff
@@ -10,6 +10,8 @@ class MangaSyncImpl : MangaSync {
 
     override var remote_id: Int = 0
 
+    override var remote_score: Float = 0f
```
Comment (inorichi) [inline app/src/main/java/eu/kanade/tachiyomi/data/database/models/MangaSyncImpl.kt]: Why do we need two scores for every sync?

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/sync/MangaSyncFragment.kt [hunk 2]
```diff
@@ -0,0 +1,509 @@
...
+        }
+    }
+
+    /**
+     * Called to open the [Manga] in browser.
+     */
+    fun openInBrowser() {
+        val source = presenter.source as? OnlineSource ?: return
+        try {
+            val url = Uri.parse(source.baseUrl + presenter.manga.url)
```
Comment (inorichi) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/sync/MangaSyncFragment.kt]: I know I wrote that, but this line should do `source.mangaDetailsRequest(manga).url().toString()` or something like that for better consistency.

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/sync/MangaSyncFragment.kt [hunk 3]
```diff
@@ -0,0 +1,509 @@
...
+    }
+
+    /**
+     * Called when [MangaSyncPresenter] has fetched [MangaSync] items
+     * When fetched update view with Sync information
+     * @param mangaSync [MangaSync] object containing information about [MangaSync].
+     */
+    fun onNextMangaSync(mangaSync: MangaSync) {
+        when (mangaSync.sync_id) {
+            MangaSyncManager.MYANIMELIST -> {
```
Comment (inorichi) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/sync/MangaSyncFragment.kt]: In my opinion the view shouldn't know about the implementations of services. The view should work with the service manager and the interface. Every custom action should be added to the interface and implemented in the services

Comment (NoodleMage) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/sync/MangaSyncFragment.kt]: I agree. Will move some functions to presenter 👍

File: app/src/main/java/eu/kanade/tachiyomi/data/database/models/MangaSyncImpl.kt (hunk 1 above)
Comment (NoodleMage) [inline app/src/main/java/eu/kanade/tachiyomi/data/database/models/MangaSyncImpl.kt]: The remote score is the average score from the service. Score is the score given by user.

File: app/src/main/java/eu/kanade/tachiyomi/ui/manga/sync/MangaSyncPresenter.kt [hunk 4]
```diff
@@ -315,6 +315,18 @@ class MangaSyncPresenter : BasePresenter<MangaSyncFragment>() {
     /**
      * TODO
      */
+    fun getStatusPosition(mangaSync: MangaSync): Int = with(context)
+    {
+        when (mangaSync.sync_id) {
```
Comment (inorichi) [inline app/src/main/java/eu/kanade/tachiyomi/ui/manga/sync/MangaSyncPresenter.kt]: These can be abstracted even more, something like `manager.getService(mangaSync.id)?.getStatusPosition(...) ?: throw ...`

File: app/src/main/java/eu/kanade/tachiyomi/data/mangasync/MangaSyncService.kt [hunk 5]
```diff
@@ -35,6 +35,48 @@ abstract class MangaSyncService(private val context: Context, val id: Int) {
...
+    abstract fun getList(): Observable<List<MangaSync>>
+
+    abstract fun search(query: String): Observable<List<MangaSync>>
+
+
+    /**
+     * Returns the spinner position of the status
+     * @return spinner position
+     */
+    abstract fun getPositionFromStatus(status: Int): Int
```
Comment (inorichi) [inline app/src/main/java/eu/kanade/tachiyomi/data/mangasync/MangaSyncService.kt]: There should be a better approach than spinner positions. I think this class should have:

`getStatusList()` which returns a list of the available status codes as an int (the value stored in the database)

`getStatus(status: Int): String` which returns the string from the code

```
// Maps a code to a string
val statusMap by lazy {
    getStatusList().associate { it to getStatus(it) }
}
```

```
// Returns the code from the string
fun getStatus(name: String): Int {
    statusMap.forEach { if (it.value == name) return it.key }
    throw Exception("Unknown status")
}
```

And for example, to display a list of status to the user, we can do: `statusMap.values`

I don't know if you are using it for something else, but IMO the services shouldn't know anything about positions.

File: app/src/main/java/eu/kanade/tachiyomi/data/database/models/MangaSyncImpl.kt (hunk 1 above)
Comment (inorichi) [inline app/src/main/java/eu/kanade/tachiyomi/data/database/models/MangaSyncImpl.kt]: But why do we need two scores for every manga? A local score is pointless if you have the other IMO (and local score would belong to `Manga`).

I'll try this weekend to do a full review, don't merge it yet.

Comment (NoodleMage) [inline app/src/main/java/eu/kanade/tachiyomi/data/database/models/MangaSyncImpl.kt]: untitled
Screenshot data is from Anilist and to lazy to make new one xD. I hope you get the gist.


The database has to be altered and the code gets unnecessary complicated if we move local score to the manga object (as far as I can think of). That's why I implemented it like this

Comment (inorichi) [inline app/src/main/java/eu/kanade/tachiyomi/data/database/models/MangaSyncImpl.kt]: Ah, now I get what the remote score is. I don't think we should support that. At most I would provide a link to open in the browser that manga in the sync service.

The more fields we support the harder it will be to maintain (that score will probably be outdated after a few days and we have another field to refresh).

Comment (NoodleMage) [inline app/src/main/java/eu/kanade/tachiyomi/data/database/models/MangaSyncImpl.kt]: I added it because it improves the layout in my opinion. But I understand where you're coming from. 

If I remove the score I will have to move it to the sync options. But I will do that some other time because I'm kinda done with this pull for a few days ;).

Comment (inorichi) [inline app/src/main/java/eu/kanade/tachiyomi/data/database/models/MangaSyncImpl.kt]: Yeah I get it, this PR is big, I can help you this weekend.

I would prefer an approach without that score, otherwise we would have more things to consider, like if a manga is synced with 2 services, would that field show the average of the two average (remote) scores?

In the end it will just give more problems, I think opening a browser tab is easier and the user can see a lot more data, but that could be in a separate PR.

**Comment by Hylak** [comment]
If feedback is welcomed, I would like to say something.

When looking for something to read the summary is the most important thing, and the new design is gonna force the scroll putting the new "buttons" and the "info" bar. So I think the bar should be removed, you already are on the info tab. 

Regarding the option buttons, in my opinion, would be better in the overflow menu because most of them are used never/once or in very few occasions (like share, sync options or open in browser). The FAB to add/remove from library is very useful as it is now, it's the most noticeable button as I think it should still be.

tl;dr: keep it simple and keep the summary and the useful data the most visible as possible.

Thanks for your work.

**Comment by inorichi** [comment]
I agree share and open in browser won't be used a lot and it would make sense to place them in the overflow menu and have more room for the relevant data (the manga's information). But IMO sync options should be more accessible (maybe an always visible button in the toolbar if the user uses sync services?). The "MAL" tab was very simple/ugly and we thought about merging it in the info tab, but maybe it wasn't the best idea...

**Comment by Hylak** [comment]
>maybe an always visible button in the toolbar if the user uses sync services?

I think that's a good solution, just 2 tabs seems fine and share/open in browser in the overflow menu.

**Comment by NoodleMage** [comment]
Will do a mock-up with your feedback when I have some energy for it xD. 

If someone with more UX experience and a feel for material design could make a mock-up would be fantastic.

**Comment by NoodleMage** [comment]
Closed because only collecting dust at the moment ;). Will reopen when there is more information about the design and the way sync gets implemented.

**Comment by NotLemon** [comment]
Any updates on this one?

**Comment by NoodleMage** [comment]
Because of the size of this pull I want to have enough free time to update the code in one go (including code review etc). Because this is only the case during the holidays I probably won't be able to update it before Christmas.

So unless someone else picks it up it will not be implemented before that time, sorry.

**Comment by inorichi** [comment]
I'll work on this one next, however I'm going to keep the third tab (I guess I'll rename it to "Tracking" unless something better is suggested).

I had the idea of showing an icon in the tab if the manga has a binding so that you can easily know its status.

**Comment by Flat** [comment]
Anything that needs to be done to move forward with this?

**Comment by inorichi** [comment]
It's almost done, it's only missing a custom layout for every score type (currently it shows a number picker from 0 to 100).

Although there's a problem. I wrote the layout with the new ConstraintLayout, and it doesn't compile in FDroid (at least when I checked it last month with other apps using it).
